### ISSUE TITLE
Better dev experience for @Inputs that expect specific strings

### ISSUE THREAD

**Issue by corykon** [HealthCatalyst/Fabric.Cashmere#1779]
For lots of cashmere components, there are inputs where devs can supply a string value that needs to match one of select list of values.

In some cases (like for button size) we have the @Input() typed as a string, and then just validate whatever is passed in:
![image](https://user-images.githubusercontent.com/12416432/148852115-a75831dc-0168-4bce-b1dc-b5305f82add1.png)

In other cases, we have the @Input() typed as a union:
![image](https://user-images.githubusercontent.com/12416432/148852290-4f01382b-6f16-4bd3-b6f9-de8f59420d1d.png)

A union-typed @Input() can cause problems for devs in some cases, depending on their typescript version and/or compiler settings. Such a case was brought up by Matt Gardner here: https://healthcatalyst.slack.com/archives/C75LHAV9S/p1641835704001300
Basically, the compiler worries that the dev is passing in a regular string and it can't be sure that the value matches the one of the unioned strings.

Devs can work around this by typing the variable they pass in with a similar unioned type. But, we should make this easier for them.

A couple options to do this:
A) Change the union-typed @Inputs to strings with validator functions, or
B) Create type aliases that are exported and become part of the library

**Comment by corykon** [comment]
I'm leaning towards the having inputs just be strings... than validating them. That feels a little more approachable. I don't want devs to worry about importing a bunch of type aliases for their component's @inputs.

**Comment by andrew-frueh** [comment]
Yeah I'd second that @corykon - and we have plenty of components that have validator functions already.  The only thing we'll need to be very careful about if we go that route is to make sure the docs clearly indicate all the acceptable values.  Because in many cases we were relying on the union type in the docs to represent that.  So we just need to make sure all the acceptable values get moved into the input's description.

**Comment by corykon** [comment]
@andrew-frueh that's a good callout, puts a little extra pressure on maintainers to keep that up to date. I think I feel ok about that

**Comment by joeskeen** [comment]
I personally think that string unions are one of the most useful and powerful features of TypeScript, and that where we are using them in Cashmere, we should continue using them. The problem here stems from what I would call "hard-coded types", where you create a type on-the-fly in the definition of the function, rather than defining the type, using it in the function declaration, and exporting it so that consumers of the library can use it as well.

Here's some examples to illustrate what I mean:

IMHO this is a sloppy API:

```ts
// you have to read the docs in order to know if you are using this right. The tools can't help you here */

private _size: string;
/** Sets size of button. Choose from: `'sm' | 'md' | 'lg'`. Defaults to `'md'`  */
@Input()
get size(): string { return this._size; }
set size(value: string) { this._size = value; }
```

This is better, in that it constrains the user's options of what string to put in. If it doesn't comply, it won't compile:

```ts
private _size: 'sm' | 'md' | 'lg';
/** Sets size of button. Choose from: `'sm' | 'md' | 'lg'`. Defaults to `'md'`  */
@Input()
get size(): 'sm' | 'md' | 'lg' { return this._size; }
set size(value: 'sm' | 'md' | 'lg' { this._size = value; }
```

But... oops!... now we are violating the DRY principle of coding. To make matters worse, since we have hard-coded this magic string union, we now cause that mess to be proliferated to the consumer of the package:

```html
<button hc-button [size]="buttonSize">CLICK ME</button>
```

```ts
class MyComponent {
  //this doesn't work:
  buttonSize: string;

  //this does, but... YUCK!
  buttonSize: 'sm' | 'md' | 'lg';
}
```

So, here is the BEST solution IMHO:

Define a type for the string union. Make it not just a haphazard collection of string unions everywhere (praying that no one will ever add, remove, or change any of them... ever), but instead give it a name, and allow it to evolve without causing undue disruption to everyone downstream! 

```ts
export type HcButtonSize = 'sm' | 'md' | 'lg';
...
export class HcButton {
  private _size: HcButtonSize;
  /** Sets size of button. Choose from: `'sm' | 'md' | 'lg'`. Defaults to `'md'`  */
  @Input()
  get size(): HcButtonSize { return this._size; }
  set size(value: HcButtonSize { this._size = value; }
}
```

For those consuming, it just works.

```html
<button hc-button [size]="buttonSize">CLICK ME</button>
```

```ts
class MyComponent {
  // simple. clean. elegant.
  buttonSize: HcButtonSize;
}
```

**Comment by corykon** [comment]
I'm agreement with everything your writing here Joe, I think those string union types are super cool and I love using them internally.

I'd maybe want to do a poll with our devs though to see how they feel about the idea. There a couple things I don't like in terms of ease of use for our devs when it comes to requiring this stuff in a component library:
1. I don't want to burden devs with having to remember to import a bunch of types everywhere they use a component. Feels like annoying boilerplate and one more thing to learn/remember
2. Often, I think devs like to drop components in their html and use hard coded strings directly in the template without the boilerplate of associated properties in the component typescript. I don't think that's a bad thing.

`<button hc-button size="lg">CLICK ME</button>`
(And double check me here... an example like this one above would have broken in Matt Garnder's case today, right?)


To take it further, imagine I have a component with say like 3 buttons in it. For each button, I might now have to create properties in my component for the buttonSize and buttonStyle for each of them.

```
class MyComponent {
  buttonOneStyle: HcButtonStyle;
  buttonTwoStyle: HcButtonStyle;
  buttonThreeStyle: HcButtonStyle;
  buttonOneSize: HcButtonSize;
  buttonTwoSize: HcButtonSize;
  buttonThreeSize: HcButtonSize;

}
```

 In UI's with lots of cashmere stuff that the dev is customizing, that boilerplate could get pretty annoying quickly. I could see dozens of properties. Maybe that's ok... I love the type safety, but worry about the usability. I really want to be careful there.

**Comment by corykon** [comment]
A note, if this syntax will still compile ok (I'm worried it won't in some cases), then I'm 100% in support of the type aliases for string union types:
```<button hc-button size="sm" buttonStyle="minimal">Nice button</button>```
